internal/graph/check_test.go

```diff
@@ -119,3 +121,203 @@ func TestCheckWithOneConcurrentGoroutineCausesNoDeadlock(t *testing.T) {
 	require.NoError(t, err)
 	require.True(t, resp.Allowed)
 }
+
+func TestCheckDatastoreQueryCount(t *testing.T) {
+	ds := memory.New()
+	defer ds.Close()
+
+	storeID := ulid.Make().String()
+
+	err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:x", "a", "user:jon"),
+		tuple.NewTupleKey("document:x", "a", "user:maria"),
+		tuple.NewTupleKey("document:x", "b", "user:maria"),
+		tuple.NewTupleKey("document:x", "parent", "org:fga"),
+		tuple.NewTupleKey("org:fga", "member", "user:maria"),
+	})
+	require.NoError(t, err)
+
+	typedefs := parser.MustParse(`
+	type user
+
+	type org
+      relations
+		define member: [user] as self
+
+	type document
+	  relations
+		define a: [user] as self
+		define b: [user] as self
+		define union as a or b
+		define union_rewrite as union
+		define intersection as a and b
+		define difference as a but not b
+		define ttu as member from parent
+        define union_and_ttu as union and ttu
+		define union_or_ttu as union or ttu or union_rewrite
+		define intersection_of_ttus as union_or_ttu and union_and_ttu
+		define parent: [org] as self
+	`)
+
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(
+		&openfgav1.AuthorizationModel{
+			Id:              ulid.Make().String(),
+			TypeDefinitions: typedefs,
+			SchemaVersion:   typesystem.SchemaVersion1_1,
+		},
+	))
+
+	tests := []struct {
+		name             string
+		check            *openfgav1.TupleKey
+		contextualTuples []*openfgav1.TupleKey
+		allowed          bool
+		minDBReads       uint32
+		maxDBReads       uint32
+	}{
+		{
+			name:       "no_direct_access",
+			check:      tuple.NewTupleKey("document:x", "a", "user:unknown"),
+			allowed:    false,
+			minDBReads: 2, // both checkDirectUserTuple and checkDirectUsersetTuples need to run
+			maxDBReads: 2,
+		},
+		{
+			name:       "direct_access",
+			check:      tuple.NewTupleKey("document:x", "a", "user:maria"),
+			allowed:    true,
+			minDBReads: 1, // checkDirectUserTuple needs to run
+			maxDBReads: 2, // maybe checkDirectUsersetTuples also runs
+		},
+		{
+			name:             "direct_access_thanks_to_contextual_tuple", // NOTE: this is counting the read from memory as a database read!
+			check:            tuple.NewTupleKey("document:x", "a", "user:unknown"),
+			contextualTuples: []*openfgav1.TupleKey{tuple.NewTupleKey("document:x", "a", "user:unknown")},
+			allowed:          true,
+			minDBReads:       1, // checkDirectUserTuple needs to run
+			maxDBReads:       2, // maybe checkDirectUsersetTuples also runs
+		},
+		{
+			name:       "union",
+			check:      tuple.NewTupleKey("document:x", "union", "user:maria"),
+			allowed:    true,
+			minDBReads: 1, // checkDirectUserTuple needs to run
+			maxDBReads: 2, // maybe checkDirectUsersetTuples also runs
```

Comment:
`maxDBReads` would be up to 4.

`document#a` can have up to 2 queries (one for checkDirectUser and one for checkDirectUserset
`document#b` can have up to 2 queries (same reason as above)

Reply:
maxDBReads here should always be the same as the worst case (e.g. you don't find a path in the evaluation tree). So this should match Line 211 below in the `union_no_access` test case.

Reply:
I see the distinction here. It wasn't clear to me.

The reason this is 2 is because with `a or b`  and `user:maria` having either `document#a` or `document#b` access, then one of them will resolve before the other, and so the max is 2.

In the cases below where `user:maria` doesn't have access this goes up to 4 because you have to evaluate `a` and `b` only to find she doesn't have access for either one.